unpack-bootstrap-tools.sh: use patchelf to undo nuke-refs and nothing else

amjoseph-nixpkgs opened the pull request:
**Note: this PR includes two commits; the first commit is #169746 which needs to merge with or before this one in order to not break CI.**  Only the second commit is specific to this PR.

The current unpack-bootstrap-tools.sh wields patchelf selectively, guided by unexplained criteria.  Specifically, it applies `patchelf --set-interpreter` to:

```
  $out/bin/* $out/libexec/gcc/*/*/*
```

(except `*/liblto*`) and `patchelf --set-interpreter --set-rpath` to:

```
  $out/lib/librt-*.so $out/lib/libpcre*
```

The provenance of these lists is unclear.  Some libraries which never had an rpath to begin with (like `librt.so`) are given an rpath by this script, while other libraries (like `libbz2.so`) are left with dangling `/nix/store/eeee...`'s.

This makes it very difficult to troubleshoot the behavior of the bootstrap-tools unpacker when porting to new platforms, where it is likely the case that patchelf is being debugged concurrently with nixpkgs.

This commit changes unpack-bootstrap-tools.sh to use patchelf *only* to replace `/nix/store/eeee...` references created by `nuke-refs`, and to replace *all* such references except where an exception is clearly described and explained.  Currently there ~~are two~~ is one such exception:

1. `libgcc_s.so` is exempted from `--set-rpath` because its rpath leaks into the final      stdenv, so we cannot patchelf its rpath without creating a      fordbidden stdenv-final->bootstrap-tools requisite.
~~2. `libpthread.so` and `libc.so` are exempted from `--set-interpreter`      because the ELF interpreter of these files does not matter (they      are libraries, not executable programs, and unlike `ld.so` they are      not meant to play both roles), and because an [unresolved bug in      patchelf](https://github.com/NixOS/patchelf/issues/368) causes corruption if it is used to set the interpreter.  Since setting the ELF interpreter of a *library* is seldom done (and indeed does not even need to be done here by nixpkgs), it is not likely that the patchelf bug will get fixed anytime soon, nor is it easy to argue for allocating resources to hunting it down.~~ Edit: no longer required (see third patch in series)

Making sure to patchelf away all of the nuke-refs'd paths will make the bootstrapping process more robust and less fragile in general.  In particular, it will prevent situations  like #169746 where the only thing preventing a bug from manifesting was the fact that the arbitrary choice of which libraries to patchelf had been made in a particular way. Situations like that make things more fragile, because now the particular choice which used to be arbitrary has become mandatory, and neither that requirement nor the reason for it is documented anywhere.


###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [X] mips64el-linux
  - [X] aarch64-linux
  - [X] powerpc64le-linux
- * `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Commit:
unpack-bootstrap-tools.sh: use patchelf to undo nuke-refs and nothing else

The current unpack-bootstrap-tools.sh wields patchelf selectively,
guided by unexplained criteria.  Specifically, it applies

patchelf --set-interpreter to:

  $out/bin/* $out/libexec/gcc/*/*/* (except */liblto*)

and patchelf --set-interpreter --set-rpath to:

  $out/lib/librt-*.so $out/lib/libpcre*

The provenance of these lists is unclear.  Some libraries which never
had an rpath to begin with (like librt.so) are given an rpath by this
script, while other libraries (like libbz2.so) are left with dangling
/nix/store/eeee...'s.

This makes it very difficult to troubleshoot the behavior of the
bootstrap-tools unpacker when porting to new platforms, where it is
likely the case that patchelf is being debugged concurrently with
nixpkgs.

This commit changes unpack-bootstrap-tools.sh to use patchelf *only*
to replace /nix/store/eeee... references created by nuke-refs, and to
replace *all* such references except where an exception is clearly
described and explained.  Currently there are two such exceptions:

  1. libgcc_s.so is exempted because its rpath leaks into the final
     stdenv, so we cannot patchelf its rpath without creating a
     fordbidden stdenv-final->bootstrap-tools requisite.

  2. libpthread.so and libc.so are exempted from --set-interpreter
     because the ELF interpreter of these files does not matter (they
     are libraries, not executable programs, and unlike ld.so they are
     not meant to play both roles), and because an unresolved bug in
     patchelf causes corruption if it is used to set the interpreter
     here:

     https://github.com/NixOS/patchelf/issues/368

Making sure to patchelf away all of the nuke-refs'd paths will make
the bootstrapping process more robust and less fragile in general.  In
particular, it will prevent situations like #169746 where the only
thing preventing a bug from manifesting was the fact that the
arbitrary choice of which libraries to patchelf had been made in a
particular way.  Situations like that make things more fragile,
because now the particular choice which used to be arbitrary has
become mandatory, and neither that requirement nor the reason for it
is documented anywhere.

This commit requires #169746 (which is included in this PR) in order
to pass CI.

a-m-joseph commented:
After cutting myself on a number of sharp edges here I'm wondering: would the maintainers accept a patch to put a statically-linked (rather than dynamically-linked) patchelf into the bootstrap-tools?  If so I will write one.

I think the unpack-bootstrap-tools.sh prior to this commit worked partly due to luck: it danced around the landmine of patchelf'ing binaries that patchelf itself linked to.  Maybe a statically linked patchelf would be worth an extra 1.7% increase in the size of bootstrap-files.tar.xz.

Commit:
stdenv/linux: remove perl enableThreading=false override from stage1

The comment for the line removed by this patch states that "A threaded
perl build needs glibc/libpthread_nonshared.a, which is not included
in bootstrapTools".  This is not the reason why it is necessary --
libpthread_nonshared.a part of the bootstrap-files on many platforms.
The reason why this workaround was needed is because nuke-refs erases
the rpath from libpthread.so, and unpack-bootstrap-tools.sh neglects
to undo this using patchelf.  Let's apply patchelf to libpthread.so,
so we aren't leaving dangling /nix/store/eeee.... refs in
libpthread.so, and remove the enableThreading=false override since it
is no longer necessary.

Additional background:

The enableThreading=false attribute not only removes the
"-Dusethreads" from the invocation of perl's Configure script, it also
patches the script like this:

  # We need to do this because the bootstrap doesn't have a static libpthread
  sed -i 's,\(libswanted.*\)pthread,\1,g' Configure

Although this prevents the perl interpreter from linking against
libpthread.so, it does not prevent the libraries bundled with the
interpreter from doing so.  Time/HiRes/HiRes.so will still try to link
against libpthread.so, and is only prevented from doing so by the fact
that unpack-bootstrap-tools.sh neglects to patchelf out all of the
nuke-refs'd rpaths.  When all of the nuke-refs'd paths are patchelf'ed
out, HiRes.so links against libpthread.so in the bootstrap tools, and
the stdenv bootstrap fails:

  Can't load '/nix/store/7ny8kmppjzx6kx4xr5kphjz8fqqrlsll-perl-5.34.1/lib/perl5/5.34.1/x86_64-linux/auto/Time/HiRes/HiRes.so' for module Time::HiRes: libpthread.so.0: cannot open shared object file: No such file or directory at /nix/store/7ny8kmppjzx6kx4xr5kphjz8fqqrlsll-perl-5.34.1/lib/perl5/5.34.1/XSLoader.pm line 93.
  at /nix/store/7ny8kmppjzx6kx4xr5kphjz8fqqrlsll-perl-5.34.1/lib/perl5/5.34.1/x86_64-linux/Time/HiRes.pm line 94.
  Compilation failed in require at lib/Autom4te/FileUtils.pm line 42.
  BEGIN failed--compilation aborted at lib/Autom4te/FileUtils.pm line 42.
  Compilation failed in require at bin/autom4te line 46.
  BEGIN failed--compilation aborted at bin/autom4te line 46.
  make[1]: *** [Makefile:2259: tests/wrapper.in] Error 2
  make[1]: *** Waiting for unfinished jobs....

Perl implements its own custom library loading routines for its
modules, which likely do not fully implement rpath searching.

This problem was revealed during the course of writing PR #161925,
which depends on this PR in order to pass CI.

a-m-joseph commented:
One other comment after having done a lot of hacking on the bootstrap-tools: it seems that the majority of the build time for the bootstrap is spent in bison's GLR parser tests.  AFAICT bison GLR isn't used by anything in the bootstrap.  The bootstrap build would run much faster with those disabled (personally I disable all the bison tests, but I can see how that might not be something that should be merged).

a-m-joseph commented:
Both the CI checks as well as my own builds on {arm64, amd64, mips64} have made it to well past the point where they're linking to or using binaries from /nix/store/*-bootstrap-tools/, so I'm marking this PR as ready for review.

a-m-joseph commented:
> The CI checks may false-fail due to the sheer amount of stuff being rebuilt leading to a timeout. That happened for my patchelf submission.

Yep: Basic evaluation checks / tests (pull_request) Cancelled after 360m — tests

a-m-joseph commented:
Changed base branch to staging.

a-m-joseph commented:
> The job running on runner GitHub Actions 4 has exceeded the maximum execution time of 360 minutes.

as expected

a-m-joseph commented:
Rebased.

SuperSandro2000 commented:
@ofborg eval

a-m-joseph commented:
This can wait until after staging closes tomorrow and reopens on 2022-May-15.  But I would like to try to move forward with it at that point; it's been pending for a pretty long time now.

Mindavi reviewed (commented):
In general this makes sense to me, but I don't feel _that_ qualified to do a proper bootstrap review. I'm not at a pc now, but I'd like to do a build too to build some confidence in this. Planning to do that later when I find some time

a-m-joseph commented on pkgs/stdenv/linux/bootstrap-tools/scripts/unpack-bootstrap-tools.sh:
```diff
@@ -21,18 +21,29 @@ fi
+    # do not --set-rpath libgcc_s.so.1, because doing so will
+    # contaminate stdenv with forbidden references to bootstrap-tools
+    if [ "x$i" == "x$out/lib/libgcc_s.so.1" ]; then continue; fi
+    # do not --set-rpath if the library had no RPATH in the first place
```
Thanks, I will update this in the morning.

a-m-joseph replied:
> Thanks, I will update this in the morning.

Sorry I am running behind here; I finally sorted out #168983 which took way longer than I thought it would.  I will get back to this PR tomorrow or Thursday.

a-m-joseph replied:
Incorporated in latest push.

a-m-joseph commented on pkgs/stdenv/linux/bootstrap-tools/scripts/unpack-bootstrap-tools.sh:
```diff
@@ -21,18 +21,29 @@ fi
         ./patchelf --set-interpreter $LD_BINARY --set-rpath $out/lib --force-rpath "$i"
 done
 
-for i in $out/lib/librt-*.so $out/lib/libpcre*; do
```
The line you highlighted was adding an RPATH to `librt.so`, which does not have an RPATH.

The unexplained reason why this worked is that every binary in the bootstrap-files is linked to `librt.so` (either directly or indirectly).  So every binary in the bootstrap-files would end up loading `librt.so`, and the RPATH from `librt.so` would therefore get included in the "aggregate RPATH" for the binary.  This is why we were getting away with leaving so many `nuke-refs`'d references to `/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-...` laying around.

If `librt.so` gets patched, and if the unstated assumption that every binary is linked to it (directly or indirectly), then I think that `patchelf`'ing `libpcre.so` is unnecessary.  I can check this in the morning if knowing would be useful.

In general the reason why the current `unpack-bootstrap-tools.sh`'s use of `patchelf` works involves a lot of unexplained reasoning like this.  It works, but it is really hard to understand *why* it works.

The main benefit of this PR is to make the reasoning really simple and straightforward: we replace all of the `/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-...` references created by `nuke-refs` (with one exception: `libgcc_s.so`, which is explained in [this comment](https://github.com/NixOS/nixpkgs/blob/a9e0e1ea21675baddaac63ee1887b9763cd39702/pkgs/stdenv/linux/bootstrap-tools/scripts/unpack-bootstrap-tools.sh#L40)), and do nothing else.

If it would help, I could rewrite the script to not only check that the `.so` has an RPATH before `patchelf`ing it, but also checks that the RPATH is `/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-...`.  This would bring us even closer to the goal of "undo what `nuke-refs` did and do nothing else".

Mindavi replied:
Thanks for the explanation, I'm starting to make sense of it a bit. I think in general this does sound like the right direction to go.

Further improvements can always be done in follow-up PRs as well.

a-m-joseph commented:
I have significantly cleaned this up and rewritten the commit message to clarify the motivation (duplicated as the top comment in this PR).

a-m-joseph commented:
Latest push just rebases on top of latest #169746; no other changes.

Commit:
remove exception for patchelf#368, as it appears to be fixed (patchelf#380)
